routing/triert/table.go

```diff
@@ -0,0 +1,227 @@
+package triert
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"sync/atomic"
+
+	tkey "github.com/libp2p/go-libp2p-xor/key"
+	"github.com/libp2p/go-libp2p-xor/trie"
+	"github.com/plprobelab/go-kademlia/key"
+	"github.com/plprobelab/go-kademlia/network/address"
+)
+
+// TODO: move to key package and replace ErrInvalidKey
+// ErrKeyWrongLength indicates a key has the wrong length
+var ErrKeyWrongLength = errors.New("key has wrong length")
+
+// TrieRT is a routing table backed by a Xor Trie which offers good scalablity and performance
+// for large networks. All exported methods are safe for concurrent use.
+type TrieRT struct {
+	self key.KadKey
+
+	// keymu is held during keys and keyNodes mutations to serialize changes
+	keymu sync.Mutex
+
+	// keys holds a pointer to an immutable trie
+	// any store to keys must be performed while holding keymu, loads may be performed without the lock
+	keys atomic.Value
```

Comment:
It allows lock free reads. Only the mutation path needs to take a lock. My assumption here is that the number of reads of a routing table is far higher than the number of mutations.

Reply:
I actually doubt that this will be the case 🤔 we're trying to add peers to the routing table after every successful `FIND_NODE`/`GET_PROVIDERS` RPC (see `query.go:L196`) and try to remove them after every failed `FIND_NODE`/`GET_PROVIDERS` RPC (see `query.go:L250`).

So we have many writes vs a single read per query. Unless we have other places where we frequently read from the routing table I believe that the assumption doesn't hold. We could batch the writes towards the end of a query which makes the #write ~= #reads.

Reply:
The current implementation makes batching writes very simple since all the mutation work is done while holding a single lock and then replacing the immutable table.

As an aside. I don't think the query should be directly mutating the routing table. It should schedule the fact that a new peer has been found to be responsive/not responsive. I could argue that this then becomes the responsibility of the refresh mechanism to probe the peer and determine whether it belongs in the routing table.

Currently the routing table is simply a map of known peer ids sorted by distance, without any liveness information. I would combine this with a refresh mechanism,  a diversity filter, a bootstrap mechanism and possibly a crawler. I don't think those functions belong in the table itself.

It's possible that we can arrange things so that we don't need concurrency control since the event queue and scheduler perform that role.

Reply:
> The current implementation makes batching writes very simple since all the mutation work is done while holding a single lock and then replacing the immutable table.

This would require a new API, right? Because the current API locks and releases for each peer addition.

We could also give `AddPeer` a variadic type parameter.

> I don't think the query should be directly mutating the routing table.

💯! This is also what I've done, in my event loop implementation.

Reply:
> This would require a new API, right? Because the current API locks and releases for each peer addition.

Yes. AddPeers or make it variadic.

Reply:
Removed concurrency support from the TrieRT